### DISCUSSION TITLE
DataPortal.Create<T> failing

### DISCUSSION THREAD

**Discussion by oskarjf** [MarimerLLC/csla#3285] [Questions] [answered]
I'm upgrading from v3.8 to v5.5.3 and for some reason my DataPortal.Create is failing.
**The failing line is:**
`DataPortal.Create<T>(new NewSheetCriteria(encId, pId, userId, userId, groupId, type, isReceivedElectronically));`

**Exception**
```
Csla.DataPortalException
  HResult=0x80131500
  Message=DataPortal.Create failed 
  Source=Csla
  StackTrace:
   at Csla.Reflection.ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError)
   at Csla.Server.DataPortal.<Create>d__10.MoveNext()
   at Csla.DataPortal`1.Create(Type objectType, Object criteria)
   at Csla.DataPortal`1.Create(Object[] criteria)
```

**The NewSheetCriteria class is defined as:**
```
        [Serializable()]
        protected class NewSheetCriteria
        {
```
with public properties and a public constructor that takes the parameters specified in DataPortal.Create.

Any ideas?

**Comment by kcabral817** [comment]
I am not sure, but I would suggest putting the parameters into a class and then passing the class as the parameter using the "criteria" class type.

**Comment by TheCakeMonster** [comment]
There are possibly two different issues to be addressed:

1. CSLA no longer looks for data access methods by name; instead, it requires them to have an attribute that depicts their usage, such as [Create], [Fetch] and so on. At a guess, the method is not found because it doesn't have the [Create] attribute.
2. The default serializer has changed; it now uses MobileFormatter by default. This has implications for criteria classes. Your criteria classes must implement the custom serialization interface that is used by MobileFormatter. The easiest way to do this is to inherit from CriteriaBase and change over to PropertyInfo-backed properties like you would use in business objects.

Here is an example of a working business object with a criteria class for a modern version of CSLA. In this case it's CSLA 6, but 5.5.3 is going to be largely the same, but for the nested factory class in place of static factory methods.

https://github.com/TheCakeMonster/Examples/blob/master/CSLA6/ProjectTracker/BlazorServer/V1/Objects/ProjectEdit.cs

**Reply by oskarjf** [reply]
1. This doesn't seem completely true as most of my DataPortal_xx methods still work though they don't have a attribute. Though adding the [Create] attribute to my DataPortal_Create method did help and I got rid of the Create error.

Now I get an error saying "ChildDataPortal.Upate failed" when calling FieldManager.UpdateChildren(this)
`Csla.DataPortalException
  HResult=0x80131500
  Message=ChildDataPortal.Update failed on the server
  Source=Csla
  StackTrace:
   at Csla.Reflection.ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError)
   at Csla.Reflection.LateBoundObject.<CallMethodTryAsyncDI>d__12`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.DataPortalTarget.<UpdateChildAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Csla.Server.ChildDataPortal.<Update>d__18.MoveNext()
   at Csla.Server.ChildDataPortal.Update(Object obj, Object[] parameters)
   at Csla.Core.FieldManager.FieldDataManager.UpdateChildren(Object[] parameters)`

`Inner Exception 3:
TargetParameterCountException: Csla.Core.BusinessBase.[InsertChild](Int32)

Inner Exception 4:
TargetParameterCountException: Csla.Core.UndoableBase.[InsertChild](Int32)

Inner Exception 5:
TargetParameterCountException: Csla.Core.BindableBase.[InsertChild](Int32)

Inner Exception 6:
TargetParameterCountException: Csla.Core.MobileObject.[InsertChild](Int32)

Inner Exception 7:
TargetParameterCountException: System.Object.[InsertChild](Int32)
`
Is it looking for a method that I'm missing? This worked fine in v3.8

**Reply by TheCakeMonster** [reply, marked as answer]
Hmm, you might be right. It might be 6.0 where the old behaviour is no longer present. Certainly the methods were dropped from the base class in 6.0, so you can no longer override them - there is nothing to override. The attributes are the way ahead though; other methods of identification are deprecated.

The TargetParameterCount exception is a similar problem to the one you had on the root object, but on a child class. It can't find an Insert method with the correct signature; the number of parameters you are passing from the parent doesn't match with any discovered insert method in the child. Adding an attribute might help with discovery, or it might be a genuine mismatch between what you are passing to the data portal and what exists in the child class.

**Reply by oskarjf** [reply]
Suspected as much. I'll try to investigate further. Would be nice if the error message where a bit more descriptive though, as in which child it could't find the method on.
